src/main/kotlin/org/rust/cargo/project/settings/RustfmtProjectSettingsService.kt

```diff
@@ -0,0 +1,47 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.cargo.project.settings
+
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.components.BaseState
+import com.intellij.openapi.components.service
+import com.intellij.openapi.project.Project
+import org.jetbrains.annotations.TestOnly
+import org.rust.cargo.toolchain.RustChannel
+
+interface RustfmtProjectSettingsService {
+
+    class State: BaseState() {
+        var additionalArguments by property("") { it.isEmpty() }
+        var channel by enum(RustChannel.DEFAULT)
+        var envs by map<String, String>()
+        var useRustfmt by property(false)
+        var runRustfmtOnSave by property(false)
+
+        fun copy(): State {
+            val state = State()
+            state.copyFrom(this)
+            return state
+        }
+    }
+
+    @TestOnly
+    fun modifyTemporary(parentDisposable: Disposable, action: (State) -> Unit)
+
+    /**
+     * Returns current state of the service.
+     * Note, result is a copy of service state, so you need to set modified state back to apply changes
+     */
+    var settingsState: State
+
+    val additionalArguments: String
+    val channel: RustChannel
+    val envs: Map<String, String>
+    val useRustfmt: Boolean
+    val runRustfmtOnSave: Boolean
```

Comment:
Do we really need these properties? Maybe fields in `State` is enough?

Reply:
We don't have access to the state from the service. And unlike state fields, these properties are read-only.

Reply:
> We don't have access to the state from the service

You can just inherit `PersistentStateComponent` to have `getState` method

> And unlike state fields, these properties are read-only.

But you want to change them anyway
Actually, I'm not sure that we need to separate interface from implementation in this case